R/learner_BART_surv_bart.R

```diff
@@ -0,0 +1,212 @@
+        }
+
+        pars = self$param_set$get_values(tags = 'train')
+
+        if (pars$importance == 'prob') {
+          sort(self$model$model$varprob.mean[-1], decreasing = T)
+        } else {
+          sort(self$model$model$varcount.mean[-1], decreasing = T)
+        }
+      }
+    ),
+
+    private = list(
+      .train = function(task) {
+        pars = self$param_set$get_values(tags = "train")
+        pars$importance = NULL # not used in the train function
+
+        x.train = as.data.frame(task$data(cols = task$feature_names))
+        times   = task$truth()[,1]
+        delta   = task$truth()[,2] # delta => status
+
+        list(
+          model = invoke(
+            BART::mc.surv.bart,
+            x.train = x.train,
+            times = times,
+            delta = delta,
+            .args = pars
+          ),
+          # need these for predict
+          x.train = x.train,
+          times = times,
+          delta = delta
+        )
+      },
+
+      .predict = function(task) {
+        # get parameters with tag "predict"
+        pars = self$param_set$get_values(tags = "predict")
+
+        # get newdata and ensure same ordering in train and predict
+        x.test = as.data.frame(ordered_features(task, self))
+
+        # subset parameters to use in `surv.pre.bart`
+        pars_pre = pars[names(pars) %in% c('K', 'events', 'ztimes', 'zdelta')]
+
+        # transform data to be suitable for BART survival analysis (needs train data)
+        trans_data = invoke(
+          BART::surv.pre.bart,
+          times   = self$model$times,
+          delta   = self$model$delta,
+          x.train = self$model$x.train,
+          x.test  = x.test,
+          .args   = pars_pre
+        )
+
+        # subset parameters to use in `predict`
+        pars_pred = pars[names(pars) %in% c('mc.cores', 'nice')]
+
+        pred_fun = function() {
+          invoke(
+            predict,
+            self$model$model,
+            newdata = trans_data$tx.test,
+            .args = pars_pred
+          )
+        }
+
+        # don't print C++ generated info during prediction
+        if (pars$quiet) {
+          utils::capture.output({
+            pred = pred_fun()
+          })
+        } else {
+          pred = pred_fun()
+        }
+
+        # Number of test observations
+        N = task$nrow
+        # Number of unique times
+        K = pred$K
+        times = pred$times
+        # Number of posterior draws
+        M = nrow(pred$surv.test)
+
+        # Convert full posterior survival matrix to 3D survival array
+        # See page 34-35 in Sparapani (2021) for more details
+        surv.array = aperm(
+          array(pred$surv.test, dim = c(M, K, N), dimnames = list(NULL, times, NULL)),
+          c(3, 2, 1)
+        )
+
+        # distr => 3d survival array
+        # crank => expected mortality
+        mlr3proba::.surv_return(times = times, surv = surv.array,
+          which.curve = pars$which.curve)
```

Comment:
what happens if `which.curve` is `NULL` ?

Reply:
to avoid this you can add the `"required"` tag to the parameter

Reply:
no need, `NULL` is the default value and is the median survival probabilities, see [here](https://github.com/mlr-org/mlr3proba/blob/main/R/surv_return.R#L89), I have tested it also.

Reply:
perfect! :)